keps/0012-framework-extensions.md

```diff
@@ -0,0 +1,319 @@
+password:
+  default: password
+  description: A more detailed description of the parameter
+data:
+  default: /path/to/sample/data.sql
+  description: Storage location of sample data to load
+```
+
+#### Example Framework Extension
+
+This framework is built from the MySQL framework defined above, but adds custom plans that allow for the loading and clearing of data that is required for a particular business application, as well as a new parameter that allows for sizing the PVC that backups are stored on.
+
+```shell
+.
+├── framework.yaml
+├── params.yaml
+└── templates
+    ├── clear-data.yaml
+    ├── pvc-size-patch.yaml
+    └── load-data.yaml
+```
+
+In order to implement these changes, we need to add the plans for `load-data` and `clear-data` and update the jobs that backup and restore the data.
+
+##### framework.yaml
+
+`framework.yaml` is the base definition of a framework. It follows the following format, extracted from the MySQL example:
+
+```yaml
+extends:
+  kudo:
+    framework: "mysql"
+    version: "5.7"
+version: "5.7"
+tasks:
+  backup:
+    from: base/backup
+    patch:
+      - pvc-size-patch.yaml
+  restore:
+    from: base/restore
+    patch:
+      - pvc-size-patch.yaml
+  load-data:
+    resources:
+      - base/init.yaml
+    patches:
+      - load-data.yaml
+  clear-data:
+    from: base/init
+    patches:
+      - clear-data.yaml
+plans:
+  resize-pv:
+    steps:
+      - name: resize
+        tasks:
+          - pv
+  load:
+    steps:
+      - name: load
+        tasks:
+          - load-data
+      - name: cleanup
+        tasks:
+          - load-data
+        delete: true
+  clear:
+    steps:
+      - name: clear
+        tasks:
+          - clear-data
+      - name: cleanup
+        tasks:
+          - clear-data
+        delete: true
+```
+
+Tasks `load-data` and `clear-data` were created with the two different specifications for how to define a task. The `backup` and `restore` tasks were updated with the new patch this framework provides
+
+##### params.yaml
+
+This framework also provides a new parameter that can be used to specify unique datasources to load, and the size of the PVC that's used for backups.
+
+```yaml
+data-location:
+  default: https://s3.aws.com/bucket/data.sql
+  description: "Location of data to load into database"
+  trigger: load
+backup-pvc-size:
+  default: 1Gi
+  description: "Size of the PVC"
+  trigger: resize-pv
+```
+
+### Implementation Details/Notes/Constraints
+
+The implementation of extensions is independent of the ability to run non-KUDO defined
+frameworks, however there are some relationships that need to be considered when extending a framework that has a different implementation engine. See forthcoming [KEP 0013](keps/0013-external-specs.md)
+
+### Outstanding Questions
+
+- How are references done? Where do I look for the base framework? Installed in the cluster? Referenced by the CLI from somewhere? Contained in the dependency folder?
```

Comment:
I would love to see this figured out in the kep before we move it to implementable. So will the CLI convert this to a "normal" package so the server side does not know about extensions? Will controller somehow evaluate the base package?

Reply:
We could do some of the following:

1) Assume the base FV is present in the cluster, maybe by installing it when we install the extension.  Concerns about chaning the base in the cluster without testing any extensions.
2) Convert the base FV into part of the Extension FV so the Extension FV CRD looks exactly the same as a regular FV.  During the install process we convert `base` template names into something unique, and adjust the names in each task appropriately.   Concerns after KEP-0013 lands in being able to identify the Engine of the base and using that differently.
3) Add a `Base` field to the FrameworkVersion (type `FrameworkVersion` or `FrameworkVersionSpec` ) and have the Base FV get inserted there.  The controller can then use the Base directly.  Concerns about complexity.

Reply:
Do we allow extending from extensions? E.g. `mysql -> mysql-extended -> mysql-extended-even-more`?

Reply:
The goal is that a framework extension looks just like any other framework, so extending extensions should be possible as well.

Reply:
oh no, inheritance even across multiple levels. I have goose bumps just when thinking about it :D but that's just me, I don't like inheritance 😛